Refactor Field Handler Classes

hutcheb opened the pull request:
First attempt at refactoring the field handler class, currently troubleshooting issue with new PlcValue of method.

hutcheb committed:
Draft for writeFloat in WriteBUffer

hutcheb committed:
Add support for writing floating point.

Note knx driver uses a 16-bit floating point. The logic needs to be 
reviewed and tested.

hutcheb committed:
Add support for Big Interger write (untested)

hutcheb committed:
replace Modbus fromPLCValue function with staticSerialise functions. 
(Needs to be fixed to include changes from Data Type PR)

hutcheb committed:
Merge branch 'develop' into Bug/AddWriteBufferSupportFloatBigInt

hutcheb committed:
Moved Modbus over to use the static Serializer.
I'll do some more testing tomorrow, but seems to work with Big Ints and 
Floats

hutcheb committed:
Fixed support for ULINT writing

hutcheb committed:
Refactored the BigInteger functions

hutcheb committed:
Refactor Field Handler Classes

hutcheb committed:
Fixed issue in new PlcValues.of function.
Is able to correctly pass standard data types now.

hutcheb committed:
Add support for writing arrays to the hello world write example.

hutcheb committed:
Merge branch 'Bug/AddWriteBufferSupportFloatBigInt' into RefactorFieldHandler

hutcheb committed:
Merge branch 'Feature/Hello-World-Write-Arrays' into RefactorFieldHandler

hutcheb committed:
Implemented staticSerialize write list functions

Outstanding:-
- Check size of buffer when writing < 8 bits
- Find out why writing < 8 bits results in a byte of all zeros.
- check endianness when writing bits.

hutcheb committed:
Merge branch 'Bug/AddWriteBufferSupportFloatBigInt' into RefactorFieldHandler

hutcheb committed:
Overrode methods on PlcValue classes and fixed list serializer

hutcheb commented:
StaticSerialzer when writing booleans needs some work.

hutcheb committed:
Add Support for writing lists of booleans

hutcheb committed:
Modified non Modbus field handler classes to modbus format
Wasn't able to test other protocols.

hutcheb commented:
Bit of a mismatch of a bunch of issues:-
- Implemented WriteBuffer - Big Integer and floating point functions.
- Implemented staticserialzer support for PlcLists
- Moved Modbus to use the static serialiser so as not to duplicate serializing code
- Refactored the field handler code, esp for Modbus, no uses a class lookup within PlcValues. Other protocols haven't been tested.
- Added support within the hello world write example for writing arrays.
- Fix for bit lengths within the WriteBuffer writeXXX functions.

chrisdutz commented on plc4j/api/src/main/java/org/apache/plc4x/java/api/value/PlcBYTE.java:
```diff
@@ -169,6 +169,30 @@ public PlcBYTE(@JsonProperty("value") short value) {
+        return (value != null) && (value <= Byte.MAX_VALUE) && (value >= Byte.MIN_VALUE);
+    }
+
+    @Override
+    @JsonIgnore
+    public byte getByte() {
+        return value.byteValue();
+    }
+
```
We now have a "byte getByte()" and a "short getBYTE()" method, I think it should only be the one or the other ... I would like to opt for the "byte getByte()". But right now I'm unsure which implications that would have.

hutcheb replied:
There's no reason to have the getBYTE,getINT,getREAL,etc... they don't get used anymore. I added these in the previous PR so we can customize the method that gets called in the data-io template by using the case name.

chrisdutz commented on plc4j/api/src/main/java/org/apache/plc4x/java/api/value/PlcBYTE.java:
```diff
@@ -181,6 +205,90 @@ public short getShort() {
         return value;
     }
 
+    @Override
+    @JsonIgnore
+    public boolean isInteger() {
+        return true;
+    }
+
+    @Override
+    @JsonIgnore
+    public int getInteger() {
+        return value.intValue();
+    }
+
```
Instead of adding these for every type, wouldn't it make sense to have something like a "PlcIECBitStringValue", "PlcIECIntegerValue", ...?

chrisdutz replied:
But thinking of it ... perhaps making it generic and passing in the type would help ...

hutcheb replied:
I think there a at least two use cases for these methods. To use them when parsing the value within the dataio classes. The other is by the user when they need to parse the value to a specific type. 

A generic function to be used when parsing like getValue() would be good.

For the user the getInteger, getByte, etc.. might be good as I can't think of an easy way to pass a type to the function and have it return the data type requested without just putting the existing functions within a switch statement.

chrisdutz commented on plc4j/api/src/main/java/org/apache/plc4x/java/api/value/PlcValues.java:
```diff
@@ -355,6 +356,28 @@ public static PlcValue of(Map<String, PlcValue> map) {
+    public static PlcValue of(Object[] values, Class clazz) {
+        //Encode values to the type defined in clazz
+        try {
+            Constructor<?> constructor = clazz.getDeclaredConstructor(values[0].getClass());
+            if(values.length == 1) {
+                return ((PlcValue) constructor.newInstance(values[0]));
+            } else {
+                return PlcValues.of(Arrays.stream(values).map(value -> (constructorHelper(constructor, value))).collect(Collectors.toList()));
+            }
```
One thing I've learnt recently, is that exceptions are extremely expensive, as creating the stack trace is expensive. is this code that is likely to produce any of "InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException" in normal execution? If yes, we should probably change this and other parts where we use similar patterns to check first instead of silently catching exceptions.

hutcheb replied:
The exception would be raised if there wasn't a class (PlcBYTE, PlcINT, etc..) that matched the datatype that was passed. It shouldn't be raised in normal operation.

chrisdutz commented on plc4j/api/src/main/java/org/apache/plc4x/java/api/value/PlcBYTE.java:
```diff
@@ -169,6 +169,30 @@ public PlcBYTE(@JsonProperty("value") short value) {
+        return true;
+    }
+
+    @Override
+    @JsonIgnore
+    public boolean getBoolean() {
+        return (value != null) && !value.equals(0);
+    }
```
In general a bit-string (at least in my perception of it, represents an array of booleans. So if for example I read one byte, from the point of the API it should be a list of 8 boolean values ... I think we need to allow this somehow. Right now the only way would be to get the "byte" value and to pick that appart.

hutcheb replied:
I was having trouble with how we would do this. I was initially thinking that if we request to read a WORD we should parse the return value as a PlcList of PlcBOOLs instead of a PlcWORD but wasn't sure how to implement this easily.

But your idea of returning a list from a getBoolean method seems reasonable.

hutcheb committed:
Reverted Commit 90fc914ec3e72a91af514742fb6dc85a34e99c9f

hutcheb committed:
Removed Class Lookup in new PlcValues.of Function
Add getPlcDataType to Plcfield Interface to allow generalizing the 
fieldHandler classes.
Add getBooleanArray() to WORD type PlcValue classes.
Add fix for staticSerializer when writing lists.

hutcheb committed:
Add checking for malformed write requests Modbus
When writing arrays if the number of values didn't match the number of 
addresses a malformed packet was sent. Additional checks have been 
adedd.
Error checking the response for writes has been added so that it reports 
error codes.
